components/mediation/data-publishers/org.wso2.micro.integrator.observability/src/main/java/org/wso2/micro/integrator/obsrvability/handler/metrics/publisher/MetricHandler.java

```diff
@@ -0,0 +1,480 @@
+                    null);
+        } else if ((null != axis2MessageContext.getProperty(MetricConstants.TRANSPORT_IN_URL) &&
+                !axis2MessageContext.getProperty(MetricConstants.TRANSPORT_IN_URL).toString().
+                        contains("services"))) {
+            try {
+                if (null != ((Axis2MessageContext) synCtx).getAxis2MessageContext().
+                        getProperty(NhttpConstants.SERVICE_PREFIX)) {
+                    String servicePort = ((Axis2MessageContext) synCtx).getAxis2MessageContext().
+                            getProperty(NhttpConstants.SERVICE_PREFIX).toString();
+                    servicePort = servicePort.substring((servicePort.lastIndexOf(':') + 1),
+                            servicePort.lastIndexOf(DELIMITER));
+                    serviceInvokePort = Integer.parseInt(servicePort);
+
+                }
+
+                if ((serviceInvokePort != internalHttpApiPort)) {
+                    String context = axis2MessageContext.getProperty(MetricConstants.TRANSPORT_IN_URL).
+                            toString();
+                    String apiInvocationUrl = axis2MessageContext.getProperty(MetricConstants.SERVICE_PREFIX).
+                            toString() + context.replaceFirst(DELIMITER, EMPTY);
+                    String apiName = getApiName(context, synCtx);
+
+                    if (apiName != null) {
+                        incrementAPICount(apiName, apiInvocationUrl);
+                        startTimers(synCtx, apiName, SynapseConstants.FAIL_SAFE_MODE_API, apiInvocationUrl);
+                    }
+                }
+            } catch (Exception ex) {
+                log.error("Error in retrieving Service Invoke Port");
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean handleRequestOutFlow(MessageContext synCtx) {
+        return true;
+    }
+
+    @Override
+    public boolean handleResponseInFlow(MessageContext synCtx) {
+        return true;
+    }
+
+    @Override
+    public boolean handleResponseOutFlow(MessageContext synCtx) {
+        org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) synCtx)
+                .getAxis2MessageContext();
+
+        int portOffset = 0;
+        int internalHttpApiPort = 0;
+        int serviceInvokePort = 0;
+
+        // Get the Internal HTTP Inbound Endpoint port
+        if ((null != System.getProperty(MetricConstants.PORT_OFFSET))) {
+            portOffset = Integer.parseInt(System.getProperty(MetricConstants.PORT_OFFSET));
+            internalHttpApiPort = Integer.parseInt(synCtx.getEnvironment().getSynapseConfiguration().
+                    getProperty(MetricConstants.INTERNAL_HTTP_API_PORT));
+            internalHttpApiPort = internalHttpApiPort + portOffset;
+
+        } else {
+            log.warn("Port Offset or Internal HTTP API port is not set.");
```

Comment:
Is it mandatory to have a port offset? Also the log says Internal HTTP API port is not set. But are we checking it?

Reply:
Refactored the implementation of getting the ports. The port offset is required to get the Internal HTTPI inbound Endpoint Port which is used to run internal APIs.